File: crates/brainwires-stores/src/conversation_store.rs
use anyhow::{Context, Result};
use chrono::Utc;
use std::sync::Arc;

use brainwires_storage::databases::{
    FieldDef, FieldType, FieldValue, Filter, Record, StorageBackend, record_get,
};

const TABLE_NAME: &str = "conversations";

/// Metadata for a conversation
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ConversationMetadata {
    /// Unique conversation identifier.
    pub conversation_id: String,
    /// Optional conversation title.
    pub title: Option<String>,
    /// Model used in this conversation.
    pub model_id: Option<String>,
    /// Creation timestamp (Unix seconds).
    pub created_at: i64,
    /// Last update timestamp (Unix seconds).
    pub updated_at: i64,
    /// Number of messages in this conversation.
    pub message_count: i32,
}

fn table_schema() -> Vec<FieldDef> {
    vec![
        FieldDef::required("conversation_id", FieldType::Utf8),
        FieldDef::optional("title", FieldType::Utf8),
        FieldDef::optional("model_id", FieldType::Utf8),
        FieldDef::required("created_at", FieldType::Int64),
        FieldDef::required("updated_at", FieldType::Int64),
        FieldDef::required("message_count", FieldType::Int32),
    ]
}

fn to_record(m: &ConversationMetadata) -> Record {
    vec![
        (
            "conversation_id".into(),
            FieldValue::Utf8(Some(m.conversation_id.clone())),
        ),
        ("title".into(), FieldValue::Utf8(m.title.clone())),
        ("model_id".into(), FieldValue::Utf8(m.model_id.clone())),
        ("created_at".into(), FieldValue::Int64(Some(m.created_at))),
        ("updated_at".into(), FieldValue::Int64(Some(m.updated_at))),
        (
            "message_count".into(),
            FieldValue::Int32(Some(m.message_count)),
        ),
    ]
}

fn from_record(r: &Record) -> Result<ConversationMetadata> {
    Ok(ConversationMetadata {
        conversation_id: record_get(r, "conversation_id")
            .and_then(|v| v.as_str())
            .context("missing conversation_id")?
            .to_string(),
        title: record_get(r, "title")
            .and_then(|v| v.as_str())
            .map(String::from),
        model_id: record_get(r, "model_id")
            .and_then(|v| v.as_str())
            .map(String::from),
        created_at: record_get(r, "created_at")
            .and_then(|v| v.as_i64())
            .context("missing created_at")?,
        updated_at: record_get(r, "updated_at")
            .and_then(|v| v.as_i64())
            .context("missing updated_at")?,
        message_count: record_get(r, "message_count")
            .and_then(|v| v.as_i32())
            .context("missing message_count")?,
    })
}

/// Store for managing conversations
pub struct ConversationStore<
    B: StorageBackend = brainwires_storage::databases::lance::LanceDatabase,
> {
    backend: Arc<B>,
}

impl<B: StorageBackend> ConversationStore<B> {
    /// Create a new conversation store
    pub fn new(backend: Arc<B>) -> Self {
        Self { backend }
    }

    /// Ensure the underlying table exists.
    pub async fn ensure_table(&self) -> Result<()> {
        self.backend.ensure_table(TABLE_NAME, &table_schema()).await
    }

    /// Create a new conversation (or update if it already exists)
    pub async fn create(
        &self,
        conversation_id: String,
        title: Option<String>,
        model_id: Option<String>,
        message_count: Option<i32>,
    ) -> Result<ConversationMetadata> {
        // Check if conversation already exists - if so, just update timestamp
        if let Ok(Some(existing)) = self.get(&conversation_id).await {
            self.update(
                &conversation_id,
                title.or(existing.title.clone()),
                message_count,
            )
            .await?;
            return self
                .get(&conversation_id)
                .await?
                .context("Conversation should exist after update");
        }

        let now = Utc::now().timestamp();

        let metadata = ConversationMetadata {
            conversation_id,
            title,
            model_id,
            created_at: now,
            updated_at: now,
            message_count: message_count.unwrap_or(0),
        };

        self.backend
            .insert(TABLE_NAME, vec![to_record(&metadata)])
            .await
            .context("Failed to create conversation")?;

        Ok(metadata)
    }

    /// Get a conversation by ID
    pub async fn get(&self, conversation_id: &str) -> Result<Option<ConversationMetadata>> {
        let filter = Filter::Eq(
            "conversation_id".into(),
            FieldValue::Utf8(Some(conversation_id.to_string())),
        );
        let records = self
            .backend
            .query(TABLE_NAME, Some(&filter), Some(1))
            .await?;

        match records.first() {
            Some(r) => Ok(Some(from_record(r)?)),
            None => Ok(None),
        }
    }

    /// List all conversations, sorted by most recently updated first
    pub async fn list(&self, limit: Option<usize>) -> Result<Vec<ConversationMetadata>> {
        let records = self.backend.query(TABLE_NAME, None, None).await?;

        let mut conversations: Vec<ConversationMetadata> =
            records.iter().filter_map(|r| from_record(r).ok()).collect();

        // Sort by updated_at descending (most recent first)
        conversations.sort_by(|a, b| b.updated_at.cmp(&a.updated_at));

        if let Some(limit) = limit {
            conversations.truncate(limit);
        }

        Ok(conversations)
    }

    /// Update conversation metadata
    pub async fn update(
        &self,
        conversation_id: &str,
        title: Option<String>,
        message_count: Option<i32>,
    ) -> Result<()> {
        let current = self
            .get(conversation_id)
            .await?
            .context("Conversation not found")?;

        // Delete current
        let filter = Filter::Eq(
            "conversation_id".into(),
            FieldValue::Utf8(Some(conversation_id.to_string())),
        );
        self.backend.delete(TABLE_NAME, &filter).await?;

        // Re-insert with updated fields
        let updated = ConversationMetadata {
            conversation_id: conversation_id.to_string(),
            title: title.or(current.title),
            model_id: current.model_id,
            created_at: current.created_at,
            updated_at: Utc::now().timestamp(),
            message_count: message_count.unwrap_or(current.message_count),
        };

        self.backend
            .insert(TABLE_NAME, vec![to_record(&updated)])
            .await?;
        Ok(())
    }

    /// Delete a conversation
    pub async fn delete(&self, conversation_id: &str) -> Result<()> {
        let filter = Filter::Eq(
            "conversation_id".into(),
            FieldValue::Utf8(Some(conversation_id.to_string())),
        );
        self.backend.delete(TABLE_NAME, &filter).await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    async fn setup() -> (
        TempDir,
        ConversationStore<brainwires_storage::databases::lance::LanceDatabase>,
    ) {
        let temp = TempDir::new().unwrap();
        let db_path = temp.path().join("test.lance");

        let backend = Arc::new(
            brainwires_storage::databases::lance::LanceDatabase::new(db_path.to_str().unwrap())
                .await
                .unwrap(),
        );
        let store = ConversationStore::new(Arc::clone(&backend));
        store.ensure_table().await.unwrap();

        (temp, store)
    }

    #[tokio::test]
    async fn test_create_conversation() {
        let (_temp, store) = setup().await;

        let conv = store
            .create(
                "test-conv-1".to_string(),
                Some("Test Conversation".to_string()),
                Some("gpt-4".to_string()),
                None,
            )
            .await
            .unwrap();

        assert_eq!(conv.conversation_id, "test-conv-1");
        assert_eq!(conv.title, Some("Test Conversation".to_string()));
        assert_eq!(conv.model_id, Some("gpt-4".to_string()));
        assert_eq!(conv.message_count, 0);
    }

    #[tokio::test]
    async fn test_create_conversation_with_message_count() {
        let (_temp, store) = setup().await;

        let conv = store
            .create(
                "test-conv-1".to_string(),
                Some("Test Conversation".to_string()),
                Some("gpt-4".to_string()),
                Some(5),
            )
            .await
            .unwrap();

        assert_eq!(conv.conversation_id, "test-conv-1");
        assert_eq!(conv.message_count, 5);
    }

    #[tokio::test]
    async fn test_get_conversation() {
        let (_temp, store) = setup().await;

        store
            .create(
                "test-conv-2".to_string(),
                Some("Test".to_string()),
                None,
                None,
            )
            .await
            .unwrap();

        let conv = store.get("test-conv-2").await.unwrap();
        assert!(conv.is_some());
        assert_eq!(conv.unwrap().conversation_id, "test-conv-2");
    }

    #[tokio::test]
    async fn test_get_nonexistent() {
        let (_temp, store) = setup().await;

        let conv = store.get("nonexistent").await.unwrap();
        assert!(conv.is_none());
    }

    #[tokio::test]
    async fn test_list_conversations() {
        let (_temp, store) = setup().await;

        store
            .create("conv-1".to_string(), Some("Conv 1".to_string()), None, None)
            .await
            .unwrap();
        store
            .create("conv-2".to_string(), Some("Conv 2".to_string()), None, None)
            .await
            .unwrap();
        store
            .create("conv-3".to_string(), Some("Conv 3".to_string()), None, None)
            .await
            .unwrap();

        let convs = store.list(None).await.unwrap();
        assert_eq!(convs.len(), 3);
    }

    #[tokio::test]
    async fn test_update_conversation() {
        let (_temp, store) = setup().await;

        store
            .create(
                "conv-update".to_string(),
                Some("Original".to_string()),
                None,
                None,
            )
            .await
            .unwrap();

        store
            .update("conv-update", Some("Updated".to_string()), Some(5))
            .await
            .unwrap();

        let conv = store.get("conv-update").await.unwrap().unwrap();
        assert_eq!(conv.title, Some("Updated".to_string()));
        assert_eq!(conv.message_count, 5);
    }

    #[tokio::test]
    async fn test_delete_conversation() {
        let (_temp, store) = setup().await;

        store
            .create("conv-delete".to_string(), None, None, None)
            .await
            .unwrap();

        let conv = store.get("conv-delete").await.unwrap();
        assert!(conv.is_some());

        store.delete("conv-delete").await.unwrap();

        let conv = store.get("conv-delete").await.unwrap();
        assert!(conv.is_none());
    }
}

File: deno/packages/stores/conversation_store.ts
/**
 * Conversation Store -- persists conversation metadata.
 *
 * Equivalent to Rust's `stores/conversation_store.rs` in rullama-storage.
 * @module
 */

import type { StorageBackend } from "@rullama/storage";
import {
  type FieldDef,
  FieldTypes,
  fieldValueAsI32,
  fieldValueAsI64,
  fieldValueAsStr,
  FieldValues,
  Filters,
  optionalField,
  type Record,
  recordGet,
  requiredField,
} from "@rullama/storage";

const TABLE_NAME = "conversations";

/** Metadata for a conversation. */
export interface ConversationMetadata {
  /** Unique conversation identifier; the key every lookup and delete uses. */
  conversationId: string;
  /** Human-readable title, if one has been set. */
  title?: string;
  /** Identifier of the model the conversation runs on, if recorded at creation. */
  modelId?: string;
  /** Creation time as a Unix timestamp in seconds. */
  createdAt: number;
  /** Last-modification time as a Unix timestamp in seconds; bumped by every `update`. */
  updatedAt: number;
  /** Number of messages in the conversation, as reported by the caller (not derived from the message store). */
  messageCount: number;
}

function tableSchema(): FieldDef[] {
  return [
    requiredField("conversation_id", FieldTypes.Utf8),
    optionalField("title", FieldTypes.Utf8),
    optionalField("model_id", FieldTypes.Utf8),
    requiredField("created_at", FieldTypes.Int64),
    requiredField("updated_at", FieldTypes.Int64),
    requiredField("message_count", FieldTypes.Int32),
  ];
}

function toRecord(m: ConversationMetadata): Record {
  return [
    ["conversation_id", FieldValues.Utf8(m.conversationId)],
    ["title", FieldValues.Utf8(m.title ?? null)],
    ["model_id", FieldValues.Utf8(m.modelId ?? null)],
    ["created_at", FieldValues.Int64(m.createdAt)],
    ["updated_at", FieldValues.Int64(m.updatedAt)],
    ["message_count", FieldValues.Int32(m.messageCount)],
  ];
}

function fromRecord(r: Record): ConversationMetadata {
  const conversationId = recordGet(r, "conversation_id");
  const createdAt = recordGet(r, "created_at");
  const updatedAt = recordGet(r, "updated_at");
  const messageCount = recordGet(r, "message_count");

  if (!conversationId || !createdAt || !updatedAt || !messageCount) {
    throw new Error("Missing required fields in conversation record");
  }

  return {
    conversationId: fieldValueAsStr(conversationId)!,
    title: recordGet(r, "title")
      ? fieldValueAsStr(recordGet(r, "title")!)
      : undefined,
    modelId: recordGet(r, "model_id")
      ? fieldValueAsStr(recordGet(r, "model_id")!)
      : undefined,
    createdAt: fieldValueAsI64(createdAt)!,
    updatedAt: fieldValueAsI64(updatedAt)!,
    messageCount: fieldValueAsI32(messageCount)!,
  };
}

/** Interface for conversation store operations. */
export interface ConversationStoreI {
  /** Create the backing `conversations` table if it does not already exist. */
  ensureTable(): Promise<void>;
  /**
   * Create a conversation, or — when one with this id already exists — update
   * its `title` / `messageCount` instead (the existing `modelId` is kept).
   * @param conversationId Id of the conversation to create.
   * @param title Optional title.
   * @param modelId Optional model identifier (only stored on a fresh create).
   * @param messageCount Initial message count (default 0).
   * @returns The metadata as stored after the create or update.
   */
  create(
    conversationId: string,
    title?: string,
    modelId?: string,
    messageCount?: number,
  ): Promise<ConversationMetadata>;
  /** Look up a conversation by id; `undefined` when none has that id. */
  get(conversationId: string): Promise<ConversationMetadata | undefined>;
  /**
   * List conversations, most recently updated first.
   * @param limit Maximum number to return; omit for all.
   */
  list(limit?: number): Promise<ConversationMetadata[]>;
  /**
   * Change `title` and/or `messageCount` of an existing conversation and set
   * `updatedAt` to now; an omitted argument keeps the current value.
   * Rejects with `"Conversation not found"` when the id is unknown.
   */
  update(
    conversationId: string,
    title?: string,
    messageCount?: number,
  ): Promise<void>;
  /** Delete the conversation with the given id (no-op when absent). */
  delete(conversationId: string): Promise<void>;
}

/** Store for managing conversations. */
export class ConversationStore implements ConversationStoreI {
  /**
   * Create a store over `backend`; call {@link ensureTable} before use.
   * @param backend Storage backend that holds the `conversations` table.
   */
  constructor(private readonly backend: StorageBackend) {}

  /** Create the `conversations` table on the backend if it is missing. */
  async ensureTable(): Promise<void> {
    await this.backend.ensureTable(TABLE_NAME, tableSchema());
  }

  /**
   * Insert a new conversation stamped with the current time, or, if the id
   * already exists, delegate to {@link update} and return the re-read record.
   */
  async create(
    conversationId: string,
    title?: string,
    modelId?: string,
    messageCount?: number,
  ): Promise<ConversationMetadata> {
    // Check if conversation already exists - if so, just update
    const existing = await this.get(conversationId);
    if (existing) {
      await this.update(conversationId, title ?? existing.title, messageCount);
      const updated = await this.get(conversationId);
      if (!updated) throw new Error("Conversation should exist after update");
      return updated;
    }

    const now = Math.floor(Date.now() / 1000);
    const metadata: ConversationMetadata = {
      conversationId,
      title,
      modelId,
      createdAt: now,
      updatedAt: now,
      messageCount: messageCount ?? 0,
    };

    await this.backend.insert(TABLE_NAME, [toRecord(metadata)]);
    return metadata;
  }

  /** Query the backend for the single record whose `conversation_id` matches. */
  async get(conversationId: string): Promise<ConversationMetadata | undefined> {
    const filter = Filters.Eq(
      "conversation_id",
      FieldValues.Utf8(conversationId),
    );
    const records = await this.backend.query(TABLE_NAME, filter, 1);
    return records.length > 0 ? fromRecord(records[0]) : undefined;
  }

  /**
   * Load every conversation from the backend, sort by `updatedAt` descending,
   * then truncate to `limit` when given.
   */
  async list(limit?: number): Promise<ConversationMetadata[]> {
    const records = await this.backend.query(TABLE_NAME);
    let conversations = records.map(fromRecord);
    conversations.sort((a, b) => b.updatedAt - a.updatedAt);
    if (limit !== undefined) {
      conversations = conversations.slice(0, limit);
    }
    return conversations;
  }

  /**
   * Rewrite the conversation's record with the new `title` / `messageCount`
   * and a fresh `updatedAt`. Implemented as a delete followed by an insert,
   * so the two steps are not atomic on the backend.
   */
  async update(
    conversationId: string,
    title?: string,
    messageCount?: number,
  ): Promise<void> {
    const current = await this.get(conversationId);
    if (!current) throw new Error("Conversation not found");

    const filter = Filters.Eq(
      "conversation_id",
      FieldValues.Utf8(conversationId),
    );
    await this.backend.delete(TABLE_NAME, filter);

    const updated: ConversationMetadata = {
      conversationId,
      title: title ?? current.title,
      modelId: current.modelId,
      createdAt: current.createdAt,
      updatedAt: Math.floor(Date.now() / 1000),
      messageCount: messageCount ?? current.messageCount,
    };

    await this.backend.insert(TABLE_NAME, [toRecord(updated)]);
  }

  /** Delete every record whose `conversation_id` matches. */
  async delete(conversationId: string): Promise<void> {
    const filter = Filters.Eq(
      "conversation_id",
      FieldValues.Utf8(conversationId),
    );
    await this.backend.delete(TABLE_NAME, filter);
  }
}

/** In-memory conversation store for testing. */
export class InMemoryConversationStore implements ConversationStoreI {
  private conversations: Map<string, ConversationMetadata> = new Map();

  /** No-op: there is no table to create in memory. */
  async ensureTable(): Promise<void> {
    await Promise.resolve();
  }

  /**
   * Store a new conversation stamped with the current time, or, if the id
   * already exists, apply {@link update} and return the stored record.
   */
  async create(
    conversationId: string,
    title?: string,
    modelId?: string,
    messageCount?: number,
  ): Promise<ConversationMetadata> {
    const existing = this.conversations.get(conversationId);
    if (existing) {
      await this.update(conversationId, title ?? existing.title, messageCount);
      return this.conversations.get(conversationId)!;
    }

    const now = Math.floor(Date.now() / 1000);
    const metadata: ConversationMetadata = {
      conversationId,
      title,
      modelId,
      createdAt: now,
      updatedAt: now,
      messageCount: messageCount ?? 0,
    };
    this.conversations.set(conversationId, metadata);
    return await Promise.resolve(metadata);
  }

  /** Return the stored record for `conversationId`, if any. */
  async get(conversationId: string): Promise<ConversationMetadata | undefined> {
    return await Promise.resolve(this.conversations.get(conversationId));
  }

  /** All stored conversations sorted by `updatedAt` descending, truncated to `limit` when given. */
  async list(limit?: number): Promise<ConversationMetadata[]> {
    let result = [...this.conversations.values()];
    result.sort((a, b) => b.updatedAt - a.updatedAt);
    if (limit !== undefined) result = result.slice(0, limit);
    return await Promise.resolve(result);
  }

  /**
   * Replace the stored record with a copy carrying the new `title` /
   * `messageCount` and a fresh `updatedAt`; throws when the id is unknown.
   */
  async update(
    conversationId: string,
    title?: string,
    messageCount?: number,
  ): Promise<void> {
    const current = this.conversations.get(conversationId);
    if (!current) throw new Error("Conversation not found");

    this.conversations.set(conversationId, {
      ...current,
      title: title ?? current.title,
      messageCount: messageCount ?? current.messageCount,
      updatedAt: Math.floor(Date.now() / 1000),
    });
    await Promise.resolve();
  }

  /** Remove the conversation from the map (no-op when absent). */
  async delete(conversationId: string): Promise<void> {
    this.conversations.delete(conversationId);
    await Promise.resolve();
  }
}
